src/main/java/org/mule/service/http/impl/service/client/GrizzlyHttpClient.java

```diff
@@ -401,27 +422,69 @@ private HttpResponse sendAndWait(HttpRequest request, HttpRequestOptions options
     checkState(asyncHttpClient != null, "The client must be started before use.");
     CompletableFuture<HttpResponse> future = new CompletableFuture<>();
     try {
       AsyncHandler<Response> asyncHandler;
+      CompletableFuture<HttpResponse> auxFuture = new CompletableFuture<>();
       if (streamingEnabled) {
-        asyncHandler = new PreservingClassLoaderAsyncHandler<>(new ResponseBodyDeferringAsyncHandler(future, responseBufferSize));
+        asyncHandler =
+            new PreservingClassLoaderAsyncHandler<>(new ResponseBodyDeferringAsyncHandler(auxFuture, responseBufferSize));
       } else {
-        asyncHandler = new PreservingClassLoaderAsyncHandler<>(new ResponseAsyncHandler(future));
+        asyncHandler = new PreservingClassLoaderAsyncHandler<>(new ResponseAsyncHandler(auxFuture));
       }
 
+      auxFuture.whenComplete((response, exception) -> {
+        if (response != null) {
+          if ((response.getStatusCode() == 301 || response.getStatusCode() == 307)
+              && response.getHeaders().containsKey("location") && options.isFollowsRedirect()) {
+
+            if (currentRedirects > MAX_REDIRECTS) {
+              future.completeExceptionally(new MaxRedirectException());
+              return;
+            }
+
+            HttpRequest redirectRequest = createRedirectRequest(response, request);
+            sendAsync(redirectRequest, options, currentRedirects + 1).whenComplete((redirectResponse, redirectException) -> {
+              if (redirectResponse != null) {
+                future.complete(redirectResponse);
+              } else {
+                future.completeExceptionally(redirectException);
+              }
+            });
+          } else {
+            future.complete(response);
+          }
+        } else {
+          future.completeExceptionally(exception);
+        }
+      });
+
       asyncHttpClient.executeRequest(createGrizzlyRequest(request, options), asyncHandler);
     } catch (Exception e) {
       future.completeExceptionally(e);
     }
     return future;
   }
 
+  private HttpRequest createRedirectRequest(HttpResponse response, HttpRequest request) {
+    Uri path = create(create(request.getUri().toString()), response.getHeaders().get("location"));
+    HttpRequestBuilder httpRequestBuilder = HttpRequest.builder();
+    httpRequestBuilder.uri(path.toUrl()).method(request.getMethod())
+        .protocol(request.getProtocol()).headers(request.getHeaders())
+        .queryParams(request.getQueryParams()).entity(request.getEntity());
+    return httpRequestBuilder.build();
+  }
+
+
   protected Request createGrizzlyRequest(HttpRequest request, HttpRequestOptions options)
       throws IOException {
     RequestBuilder reqBuilder = createRequestBuilder(request, options, builder -> {
-      builder.setFollowRedirects(options.isFollowsRedirect());
+      builder.setFollowRedirects(false);
```

Comment:
Are we sure this does not affect other types of dancing. I see tht this is the only case where the setFollowRedirects is used in the builder and would be false here always.

Reply:
For example, with a 308 response.

Reply:
Maybe take into account all the cases: 
https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections
Or at least verify that something that was working before does not break now and create another ticket.